### PR TITLE
Simplify `Skip::nth` and `Skip::last` implementations

### PR DESCRIPTION
**Pull request by ollie27**
The main improvement is to make `last` no longer recursive.

### CODE REVIEW

File: src/libcore/iter/adapters/mod.rs [hunk 1]
```diff
@@ -1827,17 +1825,13 @@ where
...
-            let next = self.next();
-            if next.is_some() {
-                // recurse. n should be 0.
-                self.last().or(next)
-            } else {
-                None
+        if self.n > 0 {
+            // nth(n) skips n+1
+            if self.iter.nth(self.n - 1).is_none() {
+                return None;
```
Comment (cuviper) [inline src/libcore/iter/adapters/mod.rs]: I like using `Try` in cases like this, `self.iter.nth(self.n - 1)?;`

Comment (ollie27) [reply src/libcore/iter/adapters/mod.rs]: That would work in this case but would make the code less symmetric with the other methods of `Skip` which is why I didn't use `?` here.

**Pull request merged**